File: avl_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        self.height = 1  # Initialize height to 1

class AVLTree:
    def __init__(self):
        self.root = None

    def get_height(self, node):
        if node is None:
            return 0
        return node.height

    def get_balance(self, node):
        if node is None:
            return 0
        return self.get_height(node.left) - self.get_height(node.right)

    def right_rotate(self, y):
        x = y.left
        T2 = x.right

        # Perform rotation
        x.right = y
        y.left = T2

        # Update heights
        y.height = max(self.get_height(y.left), self.get_height(y.right)) + 1
        x.height = max(self.get_height(x.left), self.get_height(x.right)) + 1

        return x

    def left_rotate(self, x):
        y = x.right
        T2 = y.left

        # Perform rotation
        y.left = x
        x.right = T2

        # Update heights
        x.height = max(self.get_height(x.left), self.get_height(x.right)) + 1
        y.height = max(self.get_height(y.left), self.get_height(y.right)) + 1

        return y

    def update_height_and_balance(self, node):
        if node is not None:
            node.height = max(self.get_height(node.left), self.get_height(node.right)) + 1

    def insert(self, data):
        self.root = self._insert(data, self.root)
        print("Inserted:", data)
        self.print_bfs_with_bf()

    def _insert(self, data, root):
        if root is None:
            return Node(data)
        elif data < root.data:
            root.left = self._insert(data, root.left)
        else:
            root.right = self._insert(data, root.right)

        self.update_height_and_balance(root)

        balance = self.get_balance(root)

        if balance > 1 and data < root.left.data:
            return self.right_rotate(root)

        if balance < -1 and data > root.right.data:
            return self.left_rotate(root)

        if balance > 1 and data > root.left.data:
            root.left = self.left_rotate(root.left)
            return self.right_rotate(root)

        if balance < -1 and data < root.right.data:
            root.right = self.right_rotate(root.right)
            return self.left_rotate(root)

        return root

    def find_min_node(self, root):
        current = root
        while current.left is not None:
            current = current.left
        return current

    def delete(self, data):
        self.root = self._delete(data, self.root)
        print("Deleted:", data)
        self.print_bfs_with_bf()

    def _delete(self, data, root):
        if root is None:
            return root

        if data < root.data:
            root.left = self._delete(data, root.left)
        elif data > root.data:
            root.right = self._delete(data, root.right)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            else:
                temp = self.find_min_node(root.right)
                root.data = temp.data
                root.right = self._delete(temp.data, root.right)

        self.update_height_and_balance(root)

        balance = self.get_balance(root)

        if balance > 1 and self.get_balance(root.left) >= 0:
            return self.right_rotate(root)

        if balance < -1 and self.get_balance(root.right) <= 0:
            return self.left_rotate(root)

        if balance > 1 and self.get_balance(root.left) < 0:
            root.left = self.left_rotate(root.left)
            return self.right_rotate(root)

        if balance < -1 and self.get_balance(root.right) > 0:
            root.right = self.right_rotate(root.right)
            return self.left_rotate(root)

        return root

    def print_bfs_with_bf(self):
        """Prints the tree in a breadth-first manner, showing balance factors."""
        queue = []
        if self.root is not None:
            queue.append(self.root)

        while len(queue) > 0:
            current_level = []
            next_level = []
            for node in queue:
                left_height = self.get_height(node.left)
                right_height = self.get_height(node.right)
                balance_factor = left_height - right_height
                current_level.append(f"{node.data}({balance_factor})")
                if node.left is not None:
                    next_level.append(node.left)
                if node.right is not None:
                    next_level.append(node.right)
            print(" ".join(current_level))
            queue = next_level


# Initialize the tree
tree = AVLTree()

# Menu-based system
while True:
    print("\nAVL Tree Operations:")
    print("1. Insert")
    print("2. Delete")
    print("3. Print Tree with Balance Factors")
    print("4. Exit")

    choice = input("Enter your choice: ")

    if choice == '1':
        data = int(input("Enter a value to insert: "))
        tree.insert(data)
    elif choice == '2':
        data = int(input("Enter a value to delete: "))
        tree.delete(data)
    elif choice == '3':
        print("AVL Tree with Balance Factors:")
        tree.print_bfs_with_bf()
    elif choice == '4':
        print("Exiting...")
        break
    else:
        print("Invalid choice. Please try again.")

File: avl_tree.c
#include <stdio.h>
#include <stdlib.h>

// Define structure for AVL Tree Node
typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
    int height;
} Node;

// Function to get height of a node
int getHeight(Node* node) {
    if (node == NULL)
        return 0;
    return node->height;
}

// Function to get balance factor of a node
int getBalance(Node* node) {
    if (node == NULL)
        return 0;
    return getHeight(node->left) - getHeight(node->right);
}

// Function to create a new node
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        printf("Memory allocation failed.\n");
        exit(1);
    }
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    newNode->height = 1;
    return newNode;
}

// Function to perform right rotation
Node* rightRotate(Node* y) {
    Node* x = y->left;
    Node* T2 = x->right;

    // Perform rotation
    x->right = y;
    y->left = T2;

    // Update heights
    y->height = 1 + (getHeight(y->left) > getHeight(y->right) ? getHeight(y->left) : getHeight(y->right));
    x->height = 1 + (getHeight(x->left) > getHeight(x->right) ? getHeight(x->left) : getHeight(x->right));

    return x;
}

// Function to perform left rotation
Node* leftRotate(Node* x) {
    Node* y = x->right;
    Node* T2 = y->left;

    // Perform rotation
    y->left = x;
    x->right = T2;

    // Update heights
    x->height = 1 + (getHeight(x->left) > getHeight(x->right) ? getHeight(x->left) : getHeight(x->right));
    y->height = 1 + (getHeight(y->left) > getHeight(y->right) ? getHeight(y->left) : getHeight(y->right));

    return y;
}

// Function to update height and balance of a node
void updateHeightAndBalance(Node* node) {
    if (node != NULL) {
        node->height = 1 + (getHeight(node->left) > getHeight(node->right) ? getHeight(node->left) : getHeight(node->right));
    }
}

// Function to insert a node into the AVL tree
Node* insert(Node* root, int data) {
    // Perform standard BST insertion
    if (root == NULL)
        return createNode(data);

    if (data < root->data)
        root->left = insert(root->left, data);
    else if (data > root->data)
        root->right = insert(root->right, data);
    else
        return root; // Duplicate keys not allowed

    // Update height and balance of the current node
    updateHeightAndBalance(root);

    // Check balance factor and perform rotations if necessary
    int balance = getBalance(root);

    if (balance > 1 && data < root->left->data)
        return rightRotate(root);

    if (balance < -1 && data > root->right->data)
        return leftRotate(root);

    if (balance > 1 && data > root->left->data) {
        root->left = leftRotate(root->left);
        return rightRotate(root);
    }

    if (balance < -1 && data < root->right->data) {
        root->right = rightRotate(root->right);
        return leftRotate(root);
    }

    return root;
}

// Function to find the node with minimum value in a tree rooted with given node
Node* findMinNode(Node* root) {
    Node* current = root;
    while (current->left != NULL)
        current = current->left;
    return current;
}

// Function to delete a node from the AVL tree
Node* deleteNode(Node* root, int data) {
    // Perform standard BST delete
    if (root == NULL)
        return root;

    if (data < root->data)
        root->left = deleteNode(root->left, data);
    else if (data > root->data)
        root->right = deleteNode(root->right, data);
    else {
        // Node with only one child or no child
        if (root->left == NULL) {
            Node* temp = root->right;
            free(root);
            return temp;
        } else if (root->right == NULL) {
            Node* temp = root->left;
            free(root);
            return temp;
        }

        // Node with two children: Get the inorder successor (smallest in the right subtree)
        Node* temp = findMinNode(root->right);

        // Copy the inorder successor's data to this node
        root->data = temp->data;

        // Delete the inorder successor
        root->right = deleteNode(root->right, temp->data);
    }

    // If the tree had only one node then return
    if (root == NULL)
        return root;

    // Update height and balance of the current node
    updateHeightAndBalance(root);

    // Check balance factor and perform rotations if necessary
    int balance = getBalance(root);

    if (balance > 1 && getBalance(root->left) >= 0)
        return rightRotate(root);

    if (balance < -1 && getBalance(root->right) <= 0)
        return leftRotate(root);

    if (balance > 1 && getBalance(root->left) < 0) {
        root->left = leftRotate(root->left);
        return rightRotate(root);
    }

    if (balance < -1 && getBalance(root->right) > 0) {
        root->right = rightRotate(root->right);
        return leftRotate(root);
    }

    return root;
}

// Function to print the tree in BFS manner with balance factors
void printBFSWithBF(Node* root) {
    if (root == NULL)
        return;

    // Create a queue for level order traversal
    Node* queue[1000];
    int front = 0, rear = 0;
    queue[rear++] = root;

    while (front < rear) {
        int currentLevelNodes = rear - front;
        while (currentLevelNodes--) {
            Node* current = queue[front++];
            int balanceFactor = getHeight(current->left) - getHeight(current->right);
            printf("%d(%d) ", current->data, balanceFactor);

            if (current->left != NULL)
                queue[rear++] = current->left;
            if (current->right != NULL)
                queue[rear++] = current->right;
        }
        printf("\n");
    }
}

int main() {
    Node* root = NULL;

    // Menu-based system
    while (1) {
        printf("\nAVL Tree Operations:\n");
        printf("1. Insert\n");
        printf("2. Delete\n");
        printf("3. Print Tree with Balance Factors\n");
                printf("4. Exit\n");

        int choice;
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1: {
                int data;
                printf("Enter a value to insert: ");
                scanf("%d", &data);
                root = insert(root, data);
                printf("Inserted: %d\n", data);
                printBFSWithBF(root);
                break;
            }
            case 2: {
                int data;
                printf("Enter a value to delete: ");
                scanf("%d", &data);
                root = deleteNode(root, data);
                printf("Deleted: %d\n", data);
                printBFSWithBF(root);
                break;
            }
            case 3:
                printf("AVL Tree with Balance Factors:\n");
                printBFSWithBF(root);
                break;
            case 4:
                printf("Exiting...\n");
                exit(0);
            default:
                printf("Invalid choice. Please try again.\n");
        }
    }

    return 0;
}

       
